### ISSUE TITLE
Swapchain texture usage

### ISSUE THREAD

**Issue by kvark** [gpuweb/gpuweb#437]
A Swapchain in native APIs represents a queue of presentable textures. The user doesn't create or own individual textures in that queue, but they can request (at swapchain creation) specific flags usage supported by those images.

## Metal

`CAMetalLayer` configuration has [framebufferOnly](https://developer.apple.com/documentation/quartzcore/cametallayer/1478168-framebufferonly) flag:
> If the value is true (the default), the CAMetalLayer class allocates its MTLTexture objects in ways that are optimized for display purposes. However, you may not sample, read from, or write to those textures. To support sampling and pixel read/write operations on the layer’s textures (at a cost to performance), set this value to false.

This clearly indicates that restricting the swapchain to a typical usage has performance benefits.

## D3D12
[DXGI_SWAP_CHAIN_DESC](https://docs.microsoft.com/en-us/windows/win32/api/dxgi/ns-dxgi-dxgi_swap_chain_desc) contains [DXGI_USAGE](https://docs.microsoft.com/en-us/windows/win32/direct3ddxgi/dxgi-usage) bits, of which relevant are:
  - DXGI_USAGE_RENDER_TARGET_OUTPUT
  - DXGI_USAGE_SHADER_INPUT
  - DXGI_USAGE_UNORDERED_ACCESS
  - copies appear to be always supported, but their performance [depends](https://hg.mozilla.org/mozilla-central/rev/20b3e9001dbc9b745c491bb422e158e9561e2938) on the `SHADER_INPUT` flag

MSDN doesn't appear to specify if requesting these flags has any performance implications, or if some combinations are not supported on a particular platform.

However, run-time does complain when `UNORDERED_ACCESS` is requested:
>  The BufferUsage field of the swapchain description contains some DXGI_USAGE flags that are not supported

## Vulkan

[VkSwapchainCreateInfoKHR](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkSwapchainCreateInfoKHR.html) has `imageUsage` flags that allow the user to specify the supported usage for the images.

The requested flags have to be a subset of the flags returned in `VkSurfaceCapabilitiesKHR::supportedUsageFlags` by [vkGetPhysicalDeviceSurfaceCapabilitiesKHR](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/vkGetPhysicalDeviceSurfaceCapabilitiesKHR.html).

The only guaranteed usage by the specification is `VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT`.

## TL;DR

| `GPUTextureUsage` | Metal               | D3D12 | Vulkan |
| ----------------------- | ----------------- | ------- | -------- |
| `COPY_*`                 | yes, but penalizes `OUTPUT_ATTACHMENT` | yes, but slow if `SAMPLED` isn't requested | optional |
| `SAMPLED`              | yes, but penalizes `OUTPUT_ATTACHMENT` | yes | optional  |
| `STORAGE`              | yes(?), but penalizes `OUTPUT_ATTACHMENT` | no | optional |
| `OUTPUT_ATTACHMENT`| yes | yes | yes |

Note that being able to use `STORAGE` in all of the APIs requires us to be able to create a texture view with a format that actually supports storage operations, i.e. view "rgba8unorm" as "r32uint". We don't currently support this casting in WebGPU.

**Comment by kvark** [comment]
@Kangz I guess we need a different investigation then about what what a browser-backend surface texture can do.

**Comment by kainino0x** [comment]
Editors were discussing:
- Should we initially restrict the usages of canvas textures to enable using DXGI swapchains etc?
  A: Probably not, we will have fallbacks anyway, and we can also fallback based on the texture usages.
- How desynchronized/transferControlToOffscreen interact with this, e.g. promotion to DXGI swapchain.

**Comment by kainino0x** [comment]
For the latter: I think we'll continue on #1224.

I think we can close this? Browsers should be able to implement reasonable fallback paths to deal with arbitrary usages since we have to have them for non-desynchronized canvases anyway.
